orttraining/orttraining/training_ops/cpu/tensor/split.cc

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/training_ops/cpu/tensor/split.h"
+#include "core/providers/common.h"
+#include "core/util/math.h"
+#include "core/util/math_cpuonly.h"
+
+#include "gsl/gsl"
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    SplitTraining,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder()
+        .TypeConstraint("T", DataTypeImpl::AllTensorTypes()),
+    SplitTraining);
+
+Status SplitTraining::PrepareForCompute(const TensorShape& input_shape, int num_outputs, int64_t& axis, int& before_dims,
+                                        int& after_dims_including_split_axis, int& after_dims_excluding_split,
+                                        std::vector<int64_t>& split_sizes) const {
+  auto& input_dims = input_shape.GetDims();
+  const auto num_dimensions = gsl::narrow_cast<int64_t>(input_shape.NumDimensions());
+  axis = HandleNegativeAxis(axis_, num_dimensions);  // handle negative and enforce axis is valid
+  const int64_t split_dim_size = input_dims[axis];
+
+  before_dims = gsl::narrow<int>(input_shape.SizeToDimension(axis));
+  after_dims_including_split_axis = gsl::narrow<int>(input_shape.SizeFromDimension(axis));
+  after_dims_excluding_split = (axis + 1 == num_dimensions)
+                                   ? 1  // we multiply by this value so must be 1 not 0
+                                   : gsl::narrow<int>(input_shape.SizeFromDimension(axis + 1));
+
+  std::vector<int64_t> split_sizes_values(split_sizes); 
```

Comment:
why do we need to create `split_sizes_values` when we already have `split_sizes` containing the sizes

Reply:
split_sizes will be recalculated below, I only borrow this to carry the tensor input and put it in the new variable, as I can't modify split_sizes_

---
In reply to: [459762650](https://github.com/microsoft/onnxruntime/pull/4597#discussion_r459762650) [](ancestors = 459762650)